### ISSUE TITLE
Approved: Proposal: RFC-189: Support per-team alerts and on-call rotations

### ISSUE THREAD

**Issue by bobheadxi** [sourcegraph/sourcegraph#12010]
### Context

[RFC 189](https://docs.google.com/document/d/1noZf86g2QwTbFNt3XSSbMg36D-_HCptP1thMzZsgt2Q/edit) suggests that we might want engineering teams to own their own alerts. The RFC does not yet thoroughly detail this, but this would likely entail:

* Some way to automatically send alerts to appropriate teams to handle
* Some way to denote ownership of alerts (defined in package `/monitoring`)

#### Simple alert routing

https://github.com/sourcegraph/sourcegraph/pull/11832 adds site-config-based notification definitions via Prometheus Alertmanager. We can extend the `observability.alerts` to accomodate matching on a small set of labels, for example:

```diff
{
  "level": "critical",
  "notifier": {
    "type": "opsgenie",
    "apiKey": "xxx",
    "responders": [ ... ]
  },
+ "onLabels": {
+   "service": [ "git_server", "frontend" ]
+ }
}
```

This alone might be a sufficient (if tedious) way to help teams own their own alerts. It might also cause some alerts to remain unowned. We could also restrict the implied breadth of `onLabels` field from above and just routing fields be a top-level option (example in the next point)

#### Denoting ownership

Ideally, whatever we do to denote ownership should not be Cloud-specific, ie it would be unpleasant to have to generate different alerting for Cloud. An additional required field could be added to our monitoring `Observable`s to give panel a "product area" corresponding to the teams defined in https://github.com/sourcegraph/about/pull/1150, for example:

```diff
{
	Name:            "disk_space_remaining",
	Description:     "disk space remaining by instance",
+	Owner:           OwnerSearch, // "search"
	Query:           `(src_gitserver_disk_space_available / src_gitserver_disk_space_total) * 100`,
	DataMayNotExist: true,
	Critical:        Alert{LessOrEqual: 5},
	PanelOptions:    PanelOptions().LegendFormat("{{instance}}").Unit(Percentage),
	PossibleSolutions: `
		- **Provision more disk space:** Sourcegraph will begin deleting least-used repository clones at 10% disk space remaining which may result in decreased performance, users having to wait for repositories to clone, etc.
	`,
},
```

This would:

* give us useful information on who to contact for an alert when provided alerts by customers (ie via the bug report page)
   * "a lot of alerts in group 'search' are firing, ask someone in the search team"
* make it easier to maintain notifiers by being able to just use site config and the functionality developed to deploy/silence these notifications.
   * enable us to better dogfood alerts (we'll use the same templates, alert timings, etc. as customers) 

Alternative naming for this field: `Team`, `ProductArea`, `Group`

A notifier configuration with the above ownership label them might look like:

```diff
{
  "level": "critical",
  "notifier": {
    "type": "opsgenie",
    "apiKey": "xxx",
    "responders": [ ... ]
  },
+ "owners": [ "search" ]
}
```

**Comment by slimsag** [comment]
Thanks for writing this up! I will review later this week.

**Comment by pecigonzalo** [comment]
It might be good to remove this from frontend as it is already quite of if its "domain" and we should avoid in general adding more unrelated functionality to it.

It might be simpler to the routing rules as `ConfigMap` / `Volume`.

**Comment by bobheadxi** [comment]
While I agree that routing is borderline Cloud-only (hard to imagine customers that might want to route alerts like this), I'd still advocate for the functionality, since:

* we already have extensive `dotcom` configuration: https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24%404aba276+file:%5Eschema/site%5C.schema%5C.json+%22dotcom%22:+%7B:%5Bopts%5D%7D&patternType=structural
* gives us an in-band way to dogfood new alerts https://sourcegraph.slack.com/archives/CJX299FGE/p1594734261396900
* keeps us as inline as possible with how customers configure their alerting

**Comment by pecigonzalo** [comment]
I dont think the problem is that this is cloud-only or that is contains a frontend for the config file (although I would prefer to avoid creating a config abstraction) but rather how much of the implementation logic and routing functionality becomes part of the frontend service and extends its domain/scope. As long as we keep this minimal, it should not be a problem.

**Comment by bobheadxi** [comment]
The implementation won't really touch the frontend at all - the observability alerts are handled entirely by prom-wrapper, the only thing in the frontend is the configuration

**Comment by pecigonzalo** [comment]
I think having  a "notification email" or "notification endpoint" or even "message template" simple customization on the interface and default to notify that is great, but not so sure about supporting any advanced config/routing.
That can be mounted directly to Prometheus / Alertmanager and they already have a config structure for alerts/routing. I think we could be creating complexity by abstracting that into our own JSON format configured by the sourcegraph frontend interface.

**Comment by bobheadxi** [comment]
Note that one of the goals of https://github.com/sourcegraph/sourcegraph/issues/5370 is that we should use the monitoring stack we ship - any sort of mounting alertmanager config will:

* add complexity and edge cases that `prom-wrapper` has to accommodate (it writes to the same file)
* introduce deviations in what we use vs what customers use (one of the points in #5370 is to "Remove all customizations in our ConfigMap so we use the exact same ConfigMap our customers do")
    * e.g. the custom alerts that have been added over time in prometheus are a bit hard to keep track of and work with
* feels like a workaround to avoid a feature we built

I'm not sure how strictly we want to adhere to the goals in #5370, but I think routing is a very simple implementation-wise feature that will help us avoid the points above. I've detailed other reasons in https://github.com/sourcegraph/sourcegraph/issues/12010#issuecomment-658474730

> I think having a "notification email" or "notification endpoint" or even "message template" simple customization on the interface and default to notify that is great

We don't have any of these at the moment, and I'm not sure if this is desirable (ie we probably want consistent messaging about our alerting, rather than open the door for customers to get alerts with varying descriptions)

**Comment by pecigonzalo** [comment]
I dont think it goes around it. I dont think that task, and dogfooding in general require this configuration to be part of the Sourcegraph frontend interface.

- Im not suggesting to have another thing write to it, but rather to limit how many configs we take from the frontend
- I dont think it should derivate. The task says "use the exact same ConfigMap", which I agree, we should use the same config

My concern is that we will be building and admin panel for monitoring and alerting as part of our main Sourcegraph product and creating our own abstractions over well known config formats. Something like a phpMyAdmin but for monitoring and an Opsgenie all in one.

Dogfooding is more about us consuming our own alerts/config/metrics/tools as a customer would. It does not require us to have all be configured through the Sourcegraph interface.
If we deploy using Kustomize and using https://github.com/sourcegraph/deploy-sourcegraph which might include our alerting and monitoring config, using our own docs, etc. we are dogfooding.

**Comment by bobheadxi** [comment]
> My concern is that we will be building and admin panel for monitoring and alerting as part of our main Sourcegraph product and creating our own abstractions over well known config formats. Something like a phpMyAdmin but for monitoring and an Opsgenie all in one.

context: https://github.com/sourcegraph/sourcegraph/issues/11210#issuecomment-654562906

**Comment by pecigonzalo** [comment]
Robert and I had a chat about this topic and what to add to our interface and site-config:

TLDR:
- We want to bundle alerts and metrics relevant to the application performance to ensure admins can monitor and be alerted of problems
- We want to prompt/require notifications to be setup during the site onboarding
- We want to consume the same alerts and metrics for all our or deployments
- We will not add the full Alertmanager / Prometheus functionality to our config and only the subset required for simple per-team notifications (no advanced routing, etc)
- As site config is already providing this configuration, it is no possible or even more complex to setup routing separately while allowing site admins to setup notifications

Additionally, @bobheadxi will create an RFC to capture the information for this topic.

cc/ @bobheadxi feel free to modify this comment

**Comment by slimsag** [comment]
I am late on reviewing this, sorry. Aiming to review next week.

I think this issue is sufficient in place of an RFC as it is a smaller change (as long as we clearly denote it as such via `Proposal: ` in the title, and with the expectation that reviewers approve before we implement it, in the same way an RFC would be reviewed/approved.)

**Comment by slimsag** [comment]
> #### Simple alert routing
> ...
> This alone might be a sufficient (if tedious) way to help teams own their own alerts. It might also cause some alerts to remain unowned.

Since this would be per-service alerting, I think it would not be sufficient.

This is due to the fact that the frontend service in particular is a monolith and has many teams alerts inside of its dashboard. e.g. the search team will need to monitor their alerts in the frontend dashboard, the code intel team will need to monitor theirs, etc. Similarly the distribution team or cloud team should perhaps own all provisioning alerts on all dashboards (I think that's up for discussion, but I could see that being something we want to try).

**Comment by slimsag** [comment]
> #### Denoting ownership
> An additional required field could be added to our monitoring Observables to give panel a "product area" corresponding to the teams defined in sourcegraph/about#1150, for example:
> ...
> Alternative naming for this field: Team, ProductArea.

I like this a lot 👍 naming feedback: I think `Group` is not the ideal name for such a field because it is confusing with us having actual `Group`ing of `Observable`s already, I would prefer `Owner` because it then also conveys that someone owns the actual code for that alert which is another nice property to be able to convey to others on the team.

> ```
> {
>   "level": "critical",
>   "notifier": {
>     "type": "opsgenie",
>     "apiKey": "xxx",
>     "responders": [ ... ]
>   },
> + "groups": [ "search" ]
> }
> ```

I like that this allows specifying multiple groups for a single notifier, that could be quite useful in the case of having a catch-all owner etc.

LGTM / approved, thanks for the clear writeup!

**Comment by bobheadxi** [comment]
> Since this would be per-service alerting

This is per-panel/metric (see example) - for example, a specific team would own service `gitserver`'s `disk_space_remaining`
